File: schemaTypes/assessment.ts
import { defineType, defineField } from 'sanity'
import { ClipboardIcon } from '@sanity/icons'

export const assessment = defineType({
  name: 'assessment',
  title: 'Assessment / Test',
  type: 'document',
  icon: ClipboardIcon,
  fields: [
    defineField({
      name: 'title',
      type: 'string',
      validation: (rule) => rule.required(),
    }),
    defineField({
      name: 'slug',
      type: 'slug',
      options: { source: 'title' },
      validation: (rule) => rule.required(),
    }),
    defineField({
      name: 'worksheet',
      type: 'reference',
      to: [{ type: 'worksheet' }],
      description: 'Worksheet whose questions are used for this test',
    }),
    defineField({
      name: 'durationMinutes',
      title: 'Duration (minutes)',
      type: 'number',
      validation: (rule) => rule.required().min(1),
      initialValue: 30,
    }),
    defineField({
      name: 'openInSeparateWindow',
      title: 'Open in separate test window',
      type: 'boolean',
      initialValue: true,
      description: 'When true, test is attempted in a dedicated full-screen window',
    }),
    defineField({
      name: 'passingMarksPercent',
      title: 'Passing marks (%)',
      type: 'number',
      initialValue: 40,
    }),
    defineField({
      name: 'type',
      type: 'string',
      title: 'Type',
    }),
    defineField({
      name: 'questions',
      type: 'array',
      of: [{ type: 'reference', to: [{ type: 'question' }] }],
      title: 'Questions',
    }),
    defineField({
      name: 'duration',
      type: 'number',
      title: 'Duration (minutes)',
    }),
  ],
})

export const test = {
  name: 'test',
  type: 'document',
  title: 'Test',
  fields: [
    { name: 'type', type: 'string', title: 'Type' },
    { name: 'questions', type: 'array', of: [{ type: 'reference', to: [{ type: 'question' }] }], title: 'Questions' },
    { name: 'duration', type: 'number', title: 'Duration (minutes)' },
  ],
}

File: l_m_s/features/lms/feature_lms/lib/models/assessment.dart
import 'question.dart';

/// Domain model for an assessment/test (Sanity assessment or test type).
class Assessment {
  final String id;
  final String title;
  final String? description;
  final int? durationMinutes;
  final int? totalMarks;
  final double? passingMarksPercent;
  final List<Question> questions;
  final String? worksheetId;
  final String? subjectId;

  const Assessment({
    required this.id,
    required this.title,
    this.description,
    this.durationMinutes,
    this.totalMarks,
    this.passingMarksPercent,
    this.questions = const [],
    this.worksheetId,
    this.subjectId,
  });

  factory Assessment.fromMap(Map<String, dynamic> map) {
    List<Question> questions = [];
    final qList = map['questions'] ?? map['worksheet']?['questions'];
    if (qList is List) {
      for (final q in qList) {
        if (q is Map) questions.add(Question.fromMap(Map<String, dynamic>.from(q)));
      }
    }
    String? worksheetId;
    if (map['worksheet'] is Map) worksheetId = (map['worksheet']['_id'] ?? map['worksheet']['_ref'])?.toString();
    String? subjectId;
    if (map['subject'] is Map) subjectId = (map['subject']['_id'] ?? map['subject']['_ref'])?.toString();

    return Assessment(
      id: map['_id'] as String? ?? '',
      title: map['title'] as String? ?? '',
      description: map['description'] as String?,
      durationMinutes: (map['durationMinutes'] ?? map['duration'] as num?)?.toInt(),
      totalMarks: (map['totalMarks'] as num?)?.toInt(),
      passingMarksPercent: (map['passingMarksPercent'] ?? map['passingMarks'] as num?)?.toDouble(),
      questions: questions,
      worksheetId: worksheetId,
      subjectId: subjectId,
    );
  }
}
